transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollSocketChannel.java

```diff
@@ -218,11 +218,20 @@ protected boolean doConnect(SocketAddress remoteAddress, SocketAddress localAddr
 
     private final class EpollSocketChannelUnsafe extends EpollStreamUnsafe {
         @Override
-        protected Executor closeExecutor() {
+        protected Executor prepareCloseOnExecutor() {
             try {
                 // Check isOpen() first as otherwise it will throw a RuntimeException
                 // when call getSoLinger() as the fd is not valid anymore.
                 if (isOpen() && config().getSoLinger() > 0) {
+                    // We need to cancel this key of the channel so we may not end up in a eventloop spin
+                    // because we try to read or write until the actual close happens which may be later due
+                    // SO_LINGER handling.
+                    // See https://github.com/netty/netty/issues/4449
+                    try {
+                        ((EpollEventLoop) eventLoop()).remove(EpollSocketChannel.this);
+                    } catch (Exception ignore) {
+                        // ignore
```

Comment:
What exceptions can happen here? Does it make sense to log it at DEBUG level? Just curious.